### PR TITLE
Add ParticleInterfaceBase to set the particle world index for the particle plugins.

### PR DESCRIPTION
**Pull request by MFraters**
This pull request adds a ParticleInterfaceBase to set the particle world index in the particle plugins as suggested by @gassmoeller. Let me know if this was what you had in mind.

This pull request is parallel to #5920 and should be able to be review and merged independently.

### CODE REVIEW

File: include/aspect/particle/interface.h [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+#define _aspect_particle_interface_h
+
+#include "aspect/plugins.h"
+
+namespace aspect
+{
+  namespace Particle
+  {
+    /**
+     * An abstract base class defining methods
```
Comment (tjhei) [inline include/aspect/particle/interface.h:31]: Can you explain that this is the base class for all plugins related to particles including properties, generators, etc.? What is the purpose?

File: include/aspect/particle/interface.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+     *
+     * @ingroup Particle
+     */
+    class ParticleInterfaceBase : public Plugins::InterfaceBase
+    {
+      public:
+        void set_particle_world(unsigned int particle_world_index);
```
Comment (tjhei) [inline include/aspect/particle/interface.h:38]: please document

File: include/aspect/particle/interface.h [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+      private:
+        unsigned int particle_world_index;
```
Comment (tjhei) [inline include/aspect/particle/interface.h:41]: please document what this stores.

File: include/aspect/particle/property/interface.h [hunk 4]
```diff
@@ -733,6 +734,12 @@ namespace aspect
           void
           parse_parameters (ParameterHandler &prm);
 
+          /**
+           * @brief Set the particle wrold for all particle properties
```
Comment (tjhei) [inline include/aspect/particle/property/interface.h:738]:
```suggestion
           * @brief Set the particle world for all particle properties
```

**Review by tjhei — APPROVED** [review]

**Comment by tjhei** [comment]
What is your motivation for storing the index instead of pointer?

**Comment by MFraters** [comment]
Thanks for the review! I have addressed your comments.

> What is your motivation for storing the index instead of pointer?

We need the index for the parse parameters, so since indices here used anyway, it made most sense to me to store it that way. I could change it to a pointer, but that would turn the interface into a template<int dim>, and  we would lose the information on what index the plugin is if we ever need the index during the calculation. Is there an advantage to having it as a pointer?

File: include/aspect/particle/interface.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+         * @param particle_world_index The index of the particle world this plugin belongs to.
+         */
+        void set_particle_world(unsigned int particle_world_index);
+
+      private:
+        /**
+         * Stores the index to the particle world in the
+         * particle worlds vector, to which the plugin belongs.
+         */
+        unsigned int particle_world_index;
```
Comment (gassmoeller) [inline include/aspect/particle/interface.h:52]: Yes, except I do not see how you can access this index now. Either you need to introduce a `get_particle_world` function, or you need to make this variable `protected`.

File: include/aspect/particle/interpolator/interface.h [hunk 6]
```diff
@@ -21,6 +21,7 @@
 #ifndef _aspect_particle_interpolator_interface_h
 #define _aspect_particle_interpolator_interface_h
 
+#include <aspect/particle/interface.h>
 #include <aspect/plugins.h>
```
Comment (gassmoeller) [inline include/aspect/particle/interpolator/interface.h:25]: no need to keep `plugins.h` now. Here and in the other interface headers.

**Review by gassmoeller — COMMENTED** [review]
I think this goes in the right direction. It is too late for a proper review right now, but I also think storing the index is the right approach. Every plugin that reads parameters will have to know the world index (to construct the subsection it reads the parameters from), but only very few actually need a pointer to the particle world itself (and they can get it through simulator access).

**Comment by MFraters** [comment]
@gassmoeller Thanks for your initial comments, I have addressed them.

**Review by tjhei — APPROVED** [review]

File: include/aspect/particle/interface.h [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file LICENSE.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef _aspect_particle_interface_h
+#define _aspect_particle_interface_h
+
+#include "aspect/plugins.h"
```
Comment (gassmoeller) [inline include/aspect/particle/interface.h:24]:
```suggestion
#include <aspect/plugins.h>
```

File: include/aspect/particle/interface.h [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace aspect
+{
+  namespace Particle
+  {
+    /**
+     * An abstract base class defining methods and member variables
```
Comment (gassmoeller) [inline include/aspect/particle/interface.h:31]: This class is not abstract (https://en.cppreference.com/w/cpp/language/abstract_class)
```suggestion
     * A base class defining methods and member variables
```

File: include/aspect/particle/interface.h [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+     * shared along all the particle plugins. This includes the generator,
+     * integrator, interpolator and the property classes.
+     *
+     * @ingroup Particle
+     */
+    class ParticleInterfaceBase : public Plugins::InterfaceBase
+    {
+      public:
+        /**
+         * @brief Set which particle world the plugin belong to.
```
Comment (gassmoeller) [inline include/aspect/particle/interface.h:41]:
```suggestion
         * @brief Set which particle world the plugin belongs to.
```

File: include/aspect/particle/interface.h [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+        /**
+         * @brief Gets which particle world the plugin belong to.
+         */
+        unsigned int get_particle_world_index() const;
+
+      private:
+        /**
+         * Stores the index to the particle world in the
+         * particle worlds vector, to which the plugin belongs.
```
Comment (gassmoeller) [inline include/aspect/particle/interface.h:55]: Either call it the list of particle worlds, or like this:
```suggestion
         * Stores the index to the particle world, to which the plugin belongs.
```
No one should need to know how the particle worlds are stored.

File: include/aspect/particle/property/interface.h [hunk 11]
```diff
@@ -733,6 +733,12 @@ namespace aspect
           void
           parse_parameters (ParameterHandler &prm);
 
+          /**
+           * @brief Set the particle world for all particle properties
```
Comment (gassmoeller) [inline include/aspect/particle/property/interface.h:737]:
```suggestion
           * @brief Set the particle world index for all particle properties
```

File: source/particle/interface.cc [hunk 12]
```diff
@@ -0,0 +1,41 @@
+/*
+  Copyright (C) 2015 - 2022 by the authors of the ASPECT code.
```
Comment (gassmoeller) [inline source/particle/interface.cc:2]:
```suggestion
  Copyright (C) 2024 by the authors of the ASPECT code.
```

File: source/particle/interface.cc [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+namespace aspect
+{
+  namespace Particle
+  {
+    void
+    ParticleInterfaceBase::set_particle_world_index(const unsigned int particle_world_index)
+    {
+      this->particle_world_index = particle_world_index;
+    }
+
```
Comment (gassmoeller) [inline source/particle/interface.cc:33]: 3 empty lines

File: source/particle/interface.cc [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+    unsigned int
+    ParticleInterfaceBase::get_particle_world_index() const
+    {
+      return this->particle_world_index;
```
Comment (gassmoeller) [inline source/particle/interface.cc:37]:
```suggestion
      return particle_world_index;
```

File: source/particle/interface.cc [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
```
Comment (gassmoeller) [inline source/particle/interface.cc:39]:
```suggestion
```

File: source/particle/world.cc [hunk 16]
```diff
@@ -1333,27 +1333,31 @@ namespace aspect
           sim->initialize_simulator (this->get_simulator());
         generator->parse_parameters(prm);
         generator->initialize();
+        generator->set_particle_world_index(world_index);
```
Comment (gassmoeller) [inline source/particle/world.cc:1336]: This is the wrong order, the `parse_parameters` function will already need to know this index. Here and for the other functions.

**Review by gassmoeller — COMMENTED** [review]
Yes, pretty good. My remaining comments are small, afterwards this can be merged.

**Comment by MFraters** [comment]
@gassmoeller Thanks for the review and good catch for the order of the parse parameters. I have addressed your comments.

**Review by gassmoeller — APPROVED** [review]

**Pull request merged**